File: 0712-minimum-ascii-delete-sum-for-two-strings/0712-minimum-ascii-delete-sum-for-two-strings.cs
public class Solution {
    public int MinimumDeleteSum(string s1, string s2) {
        return GetLCS(s1,s2);
    }
    private int GetLCS(string s1, string s2)
    {
        int[,] dp = new int[s1.Length+1,s2.Length+1];
        int len = LCS(s1,s2,dp);
        int i = s1.Length-1;
        int j = s2.Length-1;
        int sum1 = 0;
        while(i>=0)
        {
            Console.WriteLine((int)s1[i]);
            sum1 += (int)s1[i];
            i--;
        }
        int sum2 = 0;
        while(j>=0)
        {
            sum2 += (int)s2[j];
            j--;
        }
        Console.WriteLine(sum1+" "+sum2);
        return sum1+sum2-2*len;
    }
    private int LCS(string s1, string s2, int[,] dp)
    {
        for(int i=1; i<=s1.Length;i++)
        {
            for(int j=1; j<=s2.Length; j++)
            {
                if(s1[i-1]==s2[j-1])
                {
                    dp[i,j] = (int)s1[i-1]+dp[i-1,j-1]; // storing the ascii sum of the CS
                }
                else
                {
                    dp[i,j] = Math.Max(dp[i-1,j],dp[i,j-1]); // storing the max possible ascii sum CS (longest subsequence is not needed, instead of length, we check for the ascii sum of chars)
                }
            }
        }
        return dp[s1.Length,s2.Length];
    }
    private int GetValue(string s, string lcs)
    {
        int i=0;
        int j=0;
        int ans=0;
        while(j<lcs.Length)
        {
            if(s[i] == lcs[j])
            {
                i++;
                j++;
            }
            else
            {
                // Console.WriteLine($"adding {(int)s[i]} to {ans} for {s[i]}");
                ans += (int)s[i];
                i++;
            }
        }
        while(i<s.Length)
        {
            // Console.WriteLine($"adding {(int)s[i]} to {ans} for {s[i]}");
            ans+=(int)s[i];
            i++;
        }
        return ans;
    }
}

File: 0712-minimum-ascii-delete-sum-for-two-strings/0712-minimum-ascii-delete-sum-for-two-strings.java
class Solution {
    public int minimumDeleteSum(String s1, String s2) {
        int m = s1.length();
        int n = s2.length();
        int[][] cache = new int[m+1][n+1];
        int[] pre1 = new int[m];
        int[] pre2 = new int[n];
        int idx=0;
        for(char c : s1.toCharArray())
        {
            pre1[idx] = idx > 0 ? pre1[idx-1] + (int)(c) : (int)(c);
            idx++;
        }
        idx = 0;
        for(char c : s2.toCharArray())
        {
            pre2[idx] = idx > 0 ? pre2[idx-1] + (int)(c) : (int)(c);
            idx++;
        }
        for(int i=1; i<=m; i++)
        {
            cache[i][0] = pre1[i-1];
        }
        for(int i=1; i<=n; i++)
        {
            cache[0][i] = pre2[i-1];
        }
        for(int i=1; i<= m; i++)
        {
            for(int j=1; j<=n; j++)
            {
                if(s1.charAt(i-1) == s2.charAt(j-1))
                {
                    cache[i][j] = cache[i-1][j-1];
                }
                else
                {
                    cache[i][j] = Math.min((int)(s1.charAt(i-1)) + cache[i-1][j], (int)(s2.charAt(j-1)) + cache[i][j-1]);
                }
            }
        }
        return cache[m][n];
    }
}